server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java

```diff
@@ -0,0 +1,292 @@
+                            + "any of the accepted patterns.");
+                }
+            }
+            IBP_VERSIONS.putAll(maxInterVersion);
+        }
+    }
+
+    public String shortVersion() {
+        Pattern versionPattern = Pattern.compile("^IBP_([\\d_]+)");
+        Matcher matcher = versionPattern.matcher(this.name());
+        if (matcher.find()) {
+            String withoutIV = matcher.group(1);
+            // remove any trailing underscores
+            if (withoutIV.endsWith("_")) {
+                withoutIV = withoutIV.substring(0, withoutIV.length() - 1);
+            }
+            return withoutIV.replace("_", ".");
+        } else {
+            throw new IllegalArgumentException("Metadata version: " + this.name() + " does not fit "
+                    + "the accepted pattern.");
+        }
+    }
+
+    public String version() {
+        if (this.compareTo(IBP_0_10_0_IV0) < 0) { // versions less than this do not have IV versions
+            return shortVersion();
+        } else {
+            Pattern ivPattern = Pattern.compile("^IBP_[\\d_]+IV(\\d)");
+            Matcher matcher = ivPattern.matcher(this.name());
+            if (matcher.find()) {
+                return String.format("%s-%s", shortVersion(), matcher.group(1));
+            } else {
+                throw new IllegalArgumentException("Metadata version: " + this.name() + " does not fit "
+                        + "the accepted pattern.");
+            }
+        }
+    }
+
+    /**
+     * Return an `ApiVersion` instance for `versionString`, which can be in a variety of formats (e.g. "0.8.0", "0.8.0.x",
+     * "0.10.0", "0.10.0-IV1"). `IllegalArgumentException` is thrown if `versionString` cannot be mapped to an `ApiVersion`.
+     */
+    public static MetadataVersion apply(String versionString) {
+        String[] versionSegments = versionString.split(Pattern.quote("."));
+        int numSegments = (versionString.startsWith("0.")) ? 3 : 2;
+        String key;
+        if (numSegments >= versionSegments.length) {
+            key = versionString;
+        } else {
+            key = String.join(".", Arrays.copyOfRange(versionSegments, 0, numSegments));
+        }
+        return Optional.ofNullable(IBP_VERSIONS.get(key)).orElseThrow(() ->
+                new IllegalArgumentException("Version " + versionString + "is not a valid version")
+        );
+    }
+
+    /**
+     * Return the minimum `MetadataVersion` that supports `RecordVersion`.
+     */
+    public static MetadataVersion minSupportedFor(RecordVersion recordVersion) {
+        switch (recordVersion) {
+            case V0:
+                return IBP_0_8_0;
+            case V1:
+                return IBP_0_10_0_IV0;
+            case V2:
+                return IBP_0_11_0_IV0;
+            default:
+                throw new IllegalArgumentException("Invalid message format version " + recordVersion);
+        }
+    }
+
+    public static ApiVersionsResponse apiVersionsResponse(
```

Comment:
Hmm, why do we need these here? Could these two go in `ApiVersionsResponse` or some helper class?

Reply:
Moved it to `ApiVersionsResponse` (and suppressed the related checkstyle error) - `/Users/ahuang/kafka/clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java:19:1: Disallowed import - org.apache.kafka.clients.NodeApiVersions. [ImportControl]`